### ISSUE TITLE
Unbounded lists of results without a type annotation don't have enough type information

### ISSUE THREAD

**Issue by wiktor-obrebski** [supermacro/neverthrow#226]
This is next proposition to solve problem with #224.

I feel I learned a lot about TypeScript in last week ;) Today I tried again to provide proper types for Heterogeneous `combine` function.

First, I would like to provide my new proposition, implementation of `combine` types, based on our findings in #224 and my experimetns.

```ts
import { Result, Ok, Err } from 'neverthrow';

declare function combine<T extends readonly Result<unknown, unknown>[]>(
  results: [...T]
): Result<ExtractOkTypes<T>, ExtractErrTypes<T>[number]>;

// Given a list of Results, this extracts all the different `T` types from that list
type ExtractOkTypes<T extends readonly Result<unknown, unknown>[]> = {
  [Key in keyof T]: T[Key] extends Result<unknown, unknown>
    ? ExtractOkFromUnion<T[Key]>
    : never;
};

// Given a list of Results, this extracts all the different `E` types from that list
type ExtractErrTypes<T extends readonly Result<unknown, unknown>[]> = {
  [Key in keyof T]: T[Key] extends Result<unknown, unknown>
    ? ExtractErrFromUnion<T[Key]>
    : never;
};

// need to be separated generic type to run it for every element of union T separately
type ExtractOkFromUnion<T extends Result<unknown, unknown>> = T extends Ok<
  infer V,
  unknown
> // filter out "unknown" values
  ? V extends {}
    ? V
    : never
  : never;

// need to be separated generic type to run it for every element of union T separately
type ExtractErrFromUnion<T extends Result<unknown, unknown>> = T extends Err<
  unknown,
  infer E
> // filter out "unknown" values
  ? E extends {}
    ? E
    : never
  : never;

```

**Comment by wiktor-obrebski** [comment]
And there is a pack of types tests where I tried to cover most of the cases we discussed before.

```ts
import { ok, err, Result, Ok, Err } from 'neverthrow';

// definition of atomic data
class InvalidTest {}

const ok1 = ok('my value1');
const ok2 = ok(2);
const err1 = err(new InvalidTest());

// definition of many kinds of result-related collections that can occur
const arr = [ok1, ok2];
const arrWithErr = [ok1, err1];

const arrOfRes: (Result<string, unknown> | Result<number, unknown>)[] = [
  ok1,
  ok2,
];
const arrayOfResWithErr: (
  | Result<string, unknown>
  | Result<unknown, InvalidTest>
)[] = [ok1, err1];

const tupleOfRes: [Result<string, unknown>, Result<number, unknown>] = [
  ok1,
  ok2,
];
const tupleOfResWithErr: [
  Result<string, unknown>,
  Result<unknown, InvalidTest>
] = [ok1, err1];

// actual tests
const combinedTupleOk = combine([ok1, ok2]);
// Result<[string, number], never>

const combinedTupleErr = combine([ok1, err1]);
// Result<[string, never], InvalidTest>

const combinedArrayOk = combine(arr);
// Result<(string | number)[], never>

const combinedArrayErr = combine(arrWithErr);
// Result<string[], InvalidTest>

const combinedResultsOk = combine(arrOfRes);
// Result<(string | number)[], never>

const combinedResultsErr = combine(arrayOfResWithErr);
// Result<string[], InvalidTest>

const combinedTupleResultsOk = combine(tupleOfRes);
// Result<[string, number], never>

const combinedTupleResultsErr = combine(tupleOfResWithErr);
// Result<[string, never], InvalidTest>

```

**Comment by supermacro** [comment]
Hey @wiktor-obrebski, thanks for the detailed issue. I'll have a proper read through when I get a chance this week. Hopefully tomorrow. 

But in the mean time ...

> This is next proposition to solve problem with #224.

What exactly is the "problem"? Just so that we're on the same page. 

I'm assuming we're talking about "unbounded lists of results without a type annotation don't have enough type information and therefore `combine` ends up losing the little type information it could have inferred". 

For example:

```typescript
const resultList = [
  ok(123), // Ok<number, unknown>
  err('hello') // Err<unknown, string>
] 

const combined = combine(resultList)
```

Hence `combined` will have a type of `Result<unknown[], unknown>`.

**Comment by wiktor-obrebski** [comment]
Yes. The problem are missing/invalid types in this situation and other in `4.0.1` version. 
Is not the library trying to provide proper static types for the `combine` function? If `4.0.1` solution is acceptable for you, I can let it go of course ;)

**Comment by supermacro** [comment]
I definitely want to make sure this library is usable for more than just myself!

Going to play around with these types today. I am curious to see what the return type of calling combine is on a `Result<unknown, unknown>[]`.

**Comment by supermacro** [comment]
Interesting that you can actually work on `never` types just as though they were `unknown`. Example:

```typescript
const myList: Result<unknown, unknown>[] = [
  ok(123),
  ok('hello')
]

// type of combined is Result<string[], never> because I did a runtime check on the `never` values
const combined = combine(myList).map((values) => {
  return values.map((val) => typeof val === 'string' ? val : 'yo')
})
```

**Comment by supermacro** [comment]
@wiktor-obrebski I noticed that your type definition for `combine` changes the type of the `results` argument to be a variadic tuple on `T`. Why is that?

```diff
declare function combine<T extends readonly Result<unknown, unknown>[]>(
+  results: [...T]
-  results: T
): Result<ExtractOkTypes<T>, ExtractErrTypes<T>[number]>;
```

Ideally this is avoided because variadic tuple types came out in typescript 4, and I wouldn't want `neverthrow` to require typescript 4 as a minimum, at least not yet. 

----

Also, I noticed that I'm not getting the same types as you when calling `combine`. Do you know why that may be? Example:

<img width="746" alt="Screen Shot 2021-01-19 at 11 41 08 AM" src="https://user-images.githubusercontent.com/10621548/105064899-49ac2e80-5a4b-11eb-82e4-312e5665f0e6.png">

**Comment by wiktor-obrebski** [comment]
> Also, I noticed that I'm not getting the same types as you when calling combine. Do you know why that may be? 
I would guess it's typescript version? I testing on `4.1.3`. And you?

**Comment by wiktor-obrebski** [comment]
> @wiktor-obrebski I noticed that your type definition for combine changes the type of the results argument to be a variadic tuple on T. Why is that?

> Ideally this is avoided because variadic tuple types came out in typescript 4, and I wouldn't want neverthrow to require typescript 4 as a minimum, at least not yet.

I assumed that `4.1.3` version is ok, as it is mentioned in `neverthrow` `package.json` in dev deps.

Without spreading, this is not possible (with my knowledge) to make combine works properly with tuples input of every length.
imo it's very useful, e.g.
```ts
...
const userRes: Result<User, unknown> = getUser(userId);
const legalAgeRes: Result<number, unknown> = readFromConfig(config, 'userMinimalAge');
const userHaveAccess = Result.combine([userRes, legalAgeRes])
  .map(([user, legalAge]) => user.age >= legalAge);
```

of course it's simplified example and we could use `result.isSuccess()` etc. , but in real life both function could possible return few types of exceptions, there could be more data sources etc. etc.
this solution looks elastic for me.
without tuples, we would lost information about exact types of `user, legalAge` inside map.

**Comment by supermacro** [comment]
Hey @wiktor-obrebski , sorry for the delay. I've just started a new job so I haven't had much time to groom issues here. I'll have a look this week.

**Comment by supermacro** [comment]
$200 USD bounty for whoever solves this issue. I am open to solving it under a MAJOR semver change, since we'd be depending on more recent functionality of typescript.

**Comment by supermacro** [comment]
Bounty increased to $300 USD.

**Comment by incetarik** [comment]
> Bounty increased to $300 USD.

@supermacro Hello, could you please review the [PR](https://github.com/supermacro/neverthrow/pull/418) here?

**Comment by supermacro** [comment]
### Update

This issue is resolved within [v5.1.0](https://github.com/supermacro/neverthrow/releases/tag/v5.1.0)
